Move overlapping guards to C++.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141680
* __->__ #140013
* #139555
* #139554

This PR moves the logic for computing the overlapping relations between input tensors that
share a storage instance to C++.

In summary, this PR:

- Moves both `tensors_definitely_do_not_overlap` and part of `compute_overlapping_tensors`
to C++
- Introduces a `check_overlapping` function that re-runs `compute_overlapping_tensors`,
checking that the result is consistent with what is expected
- Introduces the `StorageOverlapChecker` class
    - Keeps track of overlapping and non-overlapping tensors
    - Actually checks the overlapping relation (call `check_overlapping`) when all tensors
    are collected
- Introduces the `STORAGE_OVERLAPPING` relational guard
    - Has a reference to a `StorageOverlapChecker`
    - Stores the to-be-checked tensors in the checker, and triggers its check
- Introduces `install_storage_overlapping_guard` python function
    - Creates an instance of `StorageOverlapChecker`
    - Creates 2 instances of the `STORAGE_OVERLAPPING` guard (for overlapping and
    non-overlapping tensors), referencing the same `StorageOverlapChecker` instance

**Why is `StorageOverlapChecker` needed?**

The way `GuardManager` is implemented, we have no control over the order in which the
check methods are called, i.e. no control over the order the tensors are collected. So, we
can't easily split them in "overlapping" and non-overlapping kinds.

Instead, we create 2 instances of `STORAGE_OVERLAPPING` guard, each of which helps
collecting the tensors for one of the kinds mentioned above. They are then used in a
single `StorageOverlapChecker` instance.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Fix test.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

bdhirsh commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -3681,6 +3969,52 @@ void install_no_tensor_aliasing_guard(
+  }
+}
+
+void install_storage_overlapping_guard(
```
cc @anijain2305 for review too. There's more context on what these guards are in the PR description, but the TLDR is:

(1) they are needed in the case where we have `f(x, y, z)`, one of our inputs is mutated, and x/y/z share storages. We need to carefully check the sizes/strides/offsets of each tensor to see if they **truly** alias, or if they don't actually overlap in memory, since this affects how we functionalize. These guards are only actually necessary to use if you have a graph with input mutation, input aliasing, and dynamic shapes

(2) executing the guard in pure python is slow, so Yukio is moving it to C++

bdhirsh commented on torch/_functorch/_aot_autograd/input_output_analysis.py:
```diff
@@ -418,15 +341,11 @@ def compute_overlapping_inputs(aot_config, fwd_inputs, aliased_input_indices):
+        actual_aliased_indices = {
+            aliased_input_indices[i]
+            for i in compute_overlapping_tensors(aliased_fwd_inputs)
+        }
```
Just noticed in the below code: if there are no dynamic shapes involved (the shapes are all static), we don't actually need `StorageOverlap` guards (since our `TENSOR_MATCH` guards already guard on the exact size/stride/offset). Are we making sure to only generate these new guards if any of the tensors' shapes in question are dynamic?

ysiraichi replied:
Ah, good catch. I will guard for that, too.

bdhirsh commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -1470,7 +1470,7 @@ def non_overlapping_args(x):
         self.assertExpectedInline(
             guard_failure,
-            """0/0: ___check_overlapping(overlapping=[L['args'][1], L['args'][2]], non_overlapping=[L['args'][0]])""",
+            """0/0: check_overlapping(overlapping=[L['args'][1], L['args'][2]], non_overlapping=[L['args'][0]])""",
```
do you have any benchmarks that show what the runtime speedups are? I was thinking of along the lines of this issue (which we can hopefully close as part of landing this PR): https://github.com/pytorch/pytorch/issues/118214

bdhirsh reviewed (commented):
pending comments above / benchmarks / any review from Animesh, this looks good to me

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
Here are the performance numbers I got from running the example in the issue 1000 times. The numbers below show the lowest time found repeating the experiment 5 times.

- No guards (12e95aa4ee4): 0.87s
- Python ([#139555](https://github.com/pytorch/pytorch/pull/139555)): 12.34s
- C++ (this PR): 5.20s

Currently investigating these numbers...

-----

<details>
<summary>Benchmark Script</summary>

```python
import torch
import torch._functorch
import torch._functorch.config
import timeit

N = 1_000
S = 10
T = 1_000
R = 5

torch._functorch.config._max_aliased_inputs_with_dynamic_shapes_enabled = N // S

@torch.compile(dynamic=True)
def f(*args):
    for a in args:
        a.add_(1)
    return args[0]

x = torch.ones(N)
args = x.split(S)

secs = timeit.repeat(
    setup="f(*args)",
    stmt="f(*args)",
    number=T,
    globals=globals(),
    repeat=R,
)

print(f"Total time for running f(*tensor({N}).split({S})) {T} times:")
for i in range(R):
    print(f"> [{i}]: {secs[i]} s")
```

</details>

ysiraichi committed:
Update

[ghstack-poisoned]

bdhirsh commented:
0.87s -> 5.20s does seem quite bad 🤔 (so ~85% of the execution time is spent in guards)

ysiraichi commented:
After running `tensor.unfold(0, 20, 10)` (creating a sliding window of `size=20, step=10`), instead of `tensor.split(10)`, I got some different results:

_Note: I changed the `no-guard` commit to c4cc9f4dc31, since it introduced a fix when we had actually aliasing inputs (`torch.unfold` wouldn't work without this)._

| Perf in Seconds          | no-guard | Python |   C++ |
|--------------------------|---------:|-------:|------:|
| `tensor.split(10)`         |     1.02 |  11.46 |  5.30 |
| `tensor.unfold(0, 20, 10)` |    13.37 |  23.52 | 10.84 |

| Number of Dynamic Shape Guards         | no-guard | Python | C++ |
|--------------------------|---------:|-------:|----:|
| `tensor.split(10)`         |    10001 |    299 | 299 |
| `tensor.unfold(0, 20, 10)` |     9901 |    395 | 395 |

It was unnexpected to me that even though `tensor.unfold` generated less dynamic shapes guards, it took more time to run it on `no-guard`. I'm thinking that it's more due to some different processing we have when we have actually aliasing inputs.

Another thing is that we are actually doing more by running `_tensors_definitely_do_not_overlap` at runtime, since we actually take into account non-contigous tensors.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
After some digging, I think I found the culprit: symbolic metadata calls! Here are the new perf numbers after the fix:

| Perf in Seconds          | no-guard | Python |   C++ |
|--------------------------|---------:|-------:|------:|
| `tensor.split(10)`         |     1.02 |  11.46 |  1.24 |
| `tensor.unfold(0, 20, 10)` |    13.37 |  23.52 | 7.54 |

ysiraichi commented:
@bdhirsh Do you think we can also merge this PR with the updated perf?

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues.

[ghstack-poisoned]

ysiraichi commented:
Turns out I was running these experiments while setting `DEBUG=1`. After turning it off, compiling in `Release` mode, I got the following numbers:

| Perf in Seconds          | no-guard | Python |   C++ |
|--------------------------|---------:|-------:|------:|
| `tensor.split(10)`         |     0.67 |  3.11 |  0.55 |
| `tensor.unfold(0, 20, 10)` |    3.39 |  5.54 | 0.94 |

That said, the investigation wasn't all meaningless. Running the old C++ version, I got 0.96 (for `split`) and 1.77 (for `unfold`) seconds, which indicated slowdown over no overlap guards. The latest version, however, yields an speedup of:

| Speedup          | C++ |
|--------------------------|---------:|
| `tensor.split(10)`         |     1.21x |
| `tensor.unfold(0, 20, 10)` |    3.60x |

ysiraichi commented:
@bdhirsh Could you take a look at this PR, whenever you have some time?

bdhirsh commented:
@ysiraichi I'm a bit confused by the chart - what is "no guard" vs C++?

If C++ corresponds to "overlapping C++ guards are checked at runtime", and no guard is "unsafely run no overlapping guards", then shouldn't it be impossible for no-guard to be slower?

ysiraichi commented:
Sorry for the confusion. Here's what I mean by each of them:

- **no-guards:** Baseline. `ShapeEnv` guards only. The `viable/strict` branch I started working on.
- **Python:** Overlapping guards implemented in Python (instance of `LAMBDA_GUARD` guard)
- **C++:** Overlapping guards implemented in C++ (instance of `STORAGE_OVERLAPPING` guard)

-----

#### Runtime Speedup

Speedup we get from running everything on C++. A speedup in that setting is possible because we are comparing:

- **no-guards:** running the `ShapeEnv` lambda guard, composed by all the guards generated from running `tensors_definitely_do_not_overlap`
- **C++:** actually running `tensors_definitely_do_not_overlap` for all pairs of aliased tensors in C++

#### True Aliasing

Whenever we have true aliasing (`torch.unfold`), we further benefit from the C++ implementation, because [it's run separately at runtime](https://github.com/pytorch/pytorch/blob/8b5c26287daa9e74f6ae2e8ace778c62fac1df1c/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1149-L1153), after the guards.

#### Compile-Time Speedup

Since we are running everything (related to the overlapping guards) in C++, we should also see a speedup at compile-time. The reason being that we were running it in Python.

ysiraichi commented:
Here are some more performance data:

| `torch.split`                        | **no-guards** |  **C++** |
|:-------------------------------------|--------------:|---------:|
| **(a)** Total Runtime                    |      0.675 ms | 0.584 ms |
| **(b)** Overlapping Check (inside guard) |            -- | 0.025 ms |
| Percentage (%)                       |            -- |     4.1% |
| **(c)** Guards                           |      0.208 ms | 0.069 ms |
| Percentage (%)                       |         30.8% |    10.2% |
| **(d)** Actual Runtime (a - c)           |      0.467 ms | 0.514 ms |
| Percentage (%)                       |         69.2% |    89.8% |

| `torch.unfold`                         | **no-guards** |  **C++** |
|:---------------------------------------|--------------:|---------:|
| **(a)** Total Runtime                      |      3.492 ms | 0.987 ms |
| **(b)** Overlapping Check (inside guard)   |            -- | 0.028 ms |
| Percentage (%)                         |            -- |     2.8% |
| **(c)** Overlapping Check (inside wrapper) |      2.430 ms | 0.041 ms |
| Percentage (%)                         |         69.6% |     4.1% |
| **(d)** Guards                             |      0.196 ms |    0.078 |
| Percentage (%)                         |          5.8% |     7.9% |
| **(e)** Actual Runtime (a - c - d)         |      0.865 ms | 0.868 ms |
| Percentage (%)                         |         24.7% |    87.9% |

-----

Note that, in order to avoid measurement overhead, some measurements were taken from different runs, i.e. by taking the lowest of 5 runs. Therefore, the "Actual Runtime" might differ between versions.

bdhirsh reviewed (approved):
sounds good to me!
